### PR TITLE
reduce clusterinfo update

### PR DESCRIPTION
**Pull request by zhiweiyin318**
related issue: https://github.com/open-cluster-management/backlog/issues/12991

### CODE REVIEW

File: pkg/klusterlet/clusterinfo/clusterinfo_controller.go [hunk 1]
```diff
@@ -388,6 +398,72 @@ func (r *ClusterInfoReconciler) getDistributionInfo() (clusterv1beta1.Distributi
...
+	}
+	return false
+}
+
+func ocpDistributionInfoUpdated(old, new *clusterv1beta1.OCPDistributionInfo) bool {
+	if len(old.AvailableUpdates) != len(new.AvailableUpdates) {
+		return true
+	}
+
+	var sortedAvailableUpdates []string
```
Comment (qiujian16) [inline pkg/klusterlet/clusterinfo/clusterinfo_controller.go:414]: I think you could use  similar of this https://github.com/open-cluster-management/work/blob/main/pkg/spoke/controllers/appliedmanifestcontroller/appliedmanifestwork_controller.go#L167-L180 to sort?

Comment (zhiweiyin318) [reply pkg/klusterlet/clusterinfo/clusterinfo_controller.go]: done

File: pkg/klusterlet/clusterinfo/clusterinfo_controller.go [hunk 2]
```diff
@@ -388,6 +399,52 @@ func (r *ClusterInfoReconciler) getDistributionInfo() (clusterv1beta1.Distributi
...
+	}
+	return false
+}
+
+func ocpDistributionInfoUpdated(old, new *clusterv1beta1.OCPDistributionInfo) bool {
+	if len(old.AvailableUpdates) != len(new.AvailableUpdates) {
+		return true
+	}
+	if old.AvailableUpdates != nil && new.AvailableUpdates != nil {
+		sort.SliceStable(old.AvailableUpdates, func(i, j int) bool { return old.AvailableUpdates[i] < old.AvailableUpdates[j] })
```
Comment (qiujian16) [inline pkg/klusterlet/clusterinfo/clusterinfo_controller.go:415]: Why do we need to sort the old one? I think it  is sorted already in the last update

Comment (zhiweiyin318) [reply pkg/klusterlet/clusterinfo/clusterinfo_controller.go:415]: my understanding is that we can't guarantee that the old one is sorted  after marshal/Unmarshal. 
use nodeStatus as an example, the original issue happened when nodeStatus in clusterinfo are different from the ones got from nodeList (slice order is different) although they have been updated using the last ones got from nodeList.

Comment (qiujian16) [reply pkg/klusterlet/clusterinfo/clusterinfo_controller.go:415]: but the old one is fetched from managedclusterinfo which should be sorted before write,  right?

Comment (zhiweiyin318) [reply pkg/klusterlet/clusterinfo/clusterinfo_controller.go]: verified that the clusterinfo is not out of order. deleted the part to sort old one.

File: pkg/klusterlet/clusterinfo/clusterinfo_controller.go [hunk 3]
```diff
@@ -388,6 +399,47 @@ func (r *ClusterInfoReconciler) getDistributionInfo() (clusterv1beta1.Distributi
...
+	}
+	return false
+}
+
+func ocpDistributionInfoUpdated(old, new *clusterv1beta1.OCPDistributionInfo) bool {
+	if len(old.AvailableUpdates) != len(new.AvailableUpdates) {
+		return true
+	}
+	if len(new.AvailableUpdates) > 1 {
+		sort.SliceStable(new.AvailableUpdates, func(i, j int) bool { return new.AvailableUpdates[i] < new.AvailableUpdates[j] })
```
Comment (qiujian16) [inline pkg/klusterlet/clusterinfo/clusterinfo_controller.go:415]: how about
```go
func ocpDistributionInfoUpdated(old, new *clusterv1beta1.OCPDistributionInfo) bool {
  sort.SliceStable(new.AvailableUpdates, func(i, j int) bool { return new.AvailableUpdates[i] < new.AvailableUpdates[j] })
   sort.SliceStable(new.VersionAvailableUpdates, func(i, j int) bool {			return new.VersionAvailableUpdates[i].Version &lt; new.VersionAvailableUpdates[j].Version		})
  sort.SliceStable(new.VersionHistory, func(i, j int) bool { return new.VersionHistory[i].Version < new.VersionHistory[j].Version })
  return !equality.Semantic.DeepEqual(old, new)
}
```

Comment (zhiweiyin318) [reply pkg/klusterlet/clusterinfo/clusterinfo_controller.go]: yes, this looks like more clear. 
 I checked that sort.SliceStable can handle empty slice.
done.

File: pkg/klusterlet/clusterinfo/clusterinfo_controller_test.go [hunk 4]
```diff
@@ -499,3 +498,174 @@ func newConfigV1Client(version string, failingStatus bool) openshiftclientset.In
...
+						{
+							Image:    "quay.io/openshift-release-dev/ocp-release@sha256:4d048ae1274d11c49f9b7e70713a072315431598b2ddbb512aee4027c422fe3e",
+							State:    "Completed",
+							Verified: false,
+							Version:  "4.5.13",
+						},
+					},
+				},
+			},
+		},
```
Comment (qiujian16) [inline pkg/klusterlet/clusterinfo/clusterinfo_controller_test.go:664]: can you add another two case
1. old is nil
2.  new is nil

Comment (zhiweiyin318) [reply pkg/klusterlet/clusterinfo/clusterinfo_controller_test.go:664]: done

**Pull request merged**